### PR TITLE
[BIOX-1950] DnP Load and save protocol release/v4.0

### PR DESCRIPTION
**Pull request by VDykyi**
- Implement save/load protocol for DnP
- add ability to generate GCode from curve
- some infill and others improvements

### CODE REVIEW

File: src/VAC/VectorAnimationComplex/InfillPattern.cpp [hunk 1]
```diff
@@ -494,7 +494,7 @@ void InfillPattern::setPattern(InfillPattern::Pattern pattern)
     pattern_ = pattern;
 }
 
-bool InfillPattern::applyInfill() const
+bool InfillPattern::isApplyInfill() const
```
Comment (jesper-cellink) [inline src/VAC/VectorAnimationComplex/InfillPattern.cpp:497]: I still find the `is` prefix odd. Is that standard Qt? I can't find it in the style guide but seems to be common naming in Qt. `shouldApplyInfill` or `isApplyingInfill` make more sense grammatically at least.  But I guess it's fine either way.

**Review by jesper-cellink — APPROVED** [review]

File: src/VAC/VectorAnimationComplex/InfillPattern.cpp (hunk 1 above)
Comment (marcus-sonestedt) [reply src/VAC/VectorAnimationComplex/InfillPattern.cpp:497]: Yeah, it's just grammar. isInfillEnabled() might be better?

File: src/VAC/Global.cpp [hunk 2]
```diff
@@ -723,6 +723,7 @@ void Global::setToolMode(Global::ToolMode mode)
 
 void Global::updateStatusBarHelp()
 {
+    return;
```
Comment (marcus-sonestedt) [inline src/VAC/Global.cpp]: early return? add comment on why it's needed.

Comment (marcus-sonestedt) [reply src/VAC/Global.cpp]: Maybe something only for us/cellink, so should be configurable.

File: src/VAC/Layer.cpp [hunk 3]
```diff
@@ -21,7 +21,7 @@
 #include "XmlStreamReader.h"
 #include "XmlStreamWriter.h"
 
-Layer::Layer(NoInit_)
+Layer::Layer(NoInit_) : absoluteLayerHeight_(-1.0)
```
Comment (marcus-sonestedt) [inline src/VAC/Layer.cpp]: set in member declaration, i.e. "float absoluteLayerHeight = -1.0;"

File: src/VAC/Layer.cpp [hunk 4]
```diff
@@ -49,7 +49,8 @@ void Layer::init_(
     connect(vac_, SIGNAL(selectionChanged()), this, SIGNAL(selectionChanged()));
 }
 
-Layer::Layer(const QString & layerName, qreal layerHeight)
+Layer::Layer(const QString & layerName, qreal layerHeight) :
+    absoluteLayerHeight_(-1.0)
```
Comment (marcus-sonestedt) [inline src/VAC/Layer.cpp]: read previous, then it doesn't need to be repeated here :)

File: src/VAC/View.cpp [hunk 5]
```diff
@@ -1153,8 +1154,12 @@ void View::drawCurve(double x, double y, ShapeDrawPhase drawPhase)
...
                 vac_->addToSelection(faceCell);
                 lastDrawnCells_ << faceCell;
-                lastDrawnCells_ << vac_->instantEdges().last();
-            }
+                // An one more edge has been created after the face has created
+                // It connects the start and end vetices and we can ignore it
+                auto lastEdge = vac_->instantEdges().last();
+                lastEdge->setShapeType(ShapeType::CURVE);
+                lastEdge->setIgnored(true);
+                lastDrawnCells_ << lastEdge;            }
```
Comment (marcusl) [inline src/VAC/View.cpp:1162]: last bracked in wrong place.. run clang-format?

**Review by marcus-sonestedt — APPROVED** [review]

**Comment by marcus-sonestedt** [comment]
never mind clang-format comment (which I deleted), this is vpaint.

File: src/VAC/Layer.cpp (hunk 4 above)
Comment (VDykyi) [reply src/VAC/Layer.cpp]: done

File: src/VAC/Layer.cpp (hunk 3 above)
Comment (VDykyi) [reply src/VAC/Layer.cpp]: done

File: src/VAC/Global.cpp (hunk 2 above)
Comment (VDykyi) [reply src/VAC/Global.cpp]: Added a constant IS_USED_STATUS_BAR = false

**Pull request merged**